--- ifftw.h.old	2013-02-05 08:50:34.347507303 -0500
+++ ifftw.h	2013-02-05 08:50:44.011507796 -0500
@@ -1026,7 +1026,7 @@
 
 double X(measure_execution_time)(const planner *plnr, 
 				 plan *pln, const problem *p);
-int X(alignment_of)(R *p);
+IFFTW_EXTERN int X(alignment_of)(R *p);
 unsigned X(hash)(const char *s);
 INT X(nbuf)(INT n, INT vl, INT maxnbuf);
 int X(nbuf_redundant)(INT n, INT vl, int which, 
