--- xianyi-OpenBLAS-347dded/exports/gensymbol   2014-06-10 09:55:47.000000000 -0400
+++ openblas-v0.2.9/exports/gensymbol   2014-06-21 19:27:06.368557354 -0400
@@ -81,6 +81,8 @@
 @misc_no_underscore_objs = (
                             goto_set_num_threads,
                             openblas_get_config,
+                            gotoblas_init,
+                            gotoblas_quit,
                            );
 
 @misc_underscore_objs = (
--- xianyi-OpenBLAS-347dded/exports/dllinit.c   2014-06-10 09:55:47.000000000 -0400
+++ openblas-v0.2.9/exports/dllinit.c   2014-06-21 19:26:44.416558346 -0400
@@ -44,7 +44,7 @@
 BOOL APIENTRY DllMain(HINSTANCE hInst, DWORD reason, LPVOID reserved) {
 
   if (reason == DLL_PROCESS_ATTACH) {
-    gotoblas_init();
+    //gotoblas_init();
   }
 
   if (reason == DLL_PROCESS_DETACH) {
--- xianyi-OpenBLAS-347dded/driver/others/memory.c	2014-06-10 09:55:47.000000000 -0400
+++ openblas-v0.2.9/driver/others/memory.c	2014-06-26 19:35:51.665009918 -0400
@@ -136,7 +136,11 @@
 
 #define BITMASK(a, b, c) ((((a) >> (b)) & (c)))
 
+#ifdef OS_WINDOWS
+#define CONSTRUCTOR
+#else
 #define CONSTRUCTOR	__attribute__ ((constructor)) 
+#endif
 #define DESTRUCTOR	__attribute__ ((destructor)) 
 
 #ifdef DYNAMIC_ARCH
@@ -769,16 +773,17 @@
   tp.PrivilegeCount = 1;
   tp.Privileges[0].Attributes = SE_PRIVILEGE_ENABLED;
   
-  if (LookupPrivilegeValue(NULL, SE_LOCK_MEMORY_NAME, &tp.Privileges[0].Luid) != TRUE) return (void *) -1;
+  if (LookupPrivilegeValue(NULL, SE_LOCK_MEMORY_NAME, &tp.Privileges[0].Luid) != TRUE) goto fail;
 
-  if (AdjustTokenPrivileges(hToken, FALSE, (PTOKEN_PRIVILEGES)&tp, 0, NULL, NULL) != TRUE) return (void *) -1;
+  if (AdjustTokenPrivileges(hToken, FALSE, &tp, 0, NULL, NULL) != TRUE) goto fail;
 
   map_address  = (void *)VirtualAlloc(address,
 				      BUFFER_SIZE,
 				      MEM_LARGE_PAGES | MEM_RESERVE | MEM_COMMIT,
 				      PAGE_READWRITE);
 
-  AdjustTokenPrivileges(hToken, TRUE, &tp, 0, (PTOKEN_PRIVILEGES)NULL, NULL);
+  tp.Privileges[0].Attributes = 0;
+  AdjustTokenPrivileges(hToken, FALSE, &tp, 0, NULL, NULL);
 
   if (map_address == (void *)NULL) map_address = (void *)-1;
   
@@ -791,6 +796,10 @@
   }
 
   return map_address;
+
+fail:
+  CloseHandle(hToken);
+  return -1;
 }
 #endif
