--- Makefile.old	2014-01-23 20:06:09.000000000 -0800
+++ Makefile	2014-01-23 20:06:11.000000000 -0800
@@ -5,6 +5,7 @@
 
 cflags = -O2 -std=c99 -pedantic -Wall -fpic $(CFLAGS)
 cc = $(CC) $(cflags)
+AR = ar
 
 
 # meta targets
@@ -31,7 +32,7 @@
 
 libutf8proc.a: utf8proc.o
 	rm -f libutf8proc.a
-	ar rs libutf8proc.a utf8proc.o
+	$(AR) rs libutf8proc.a utf8proc.o
 
 libutf8proc.so: utf8proc.o
 	$(cc) -shared -o libutf8proc.so utf8proc.o
