From 7afc76b107d119cf49ee90dc022b64468f2581c2 Mon Sep 17 00:00:00 2001
From: Tony Kelman <tony@kelman.net>
Date: Sun, 11 Sep 2016 21:16:30 -0700
Subject: [PATCH] Only define USE_NTLM with USE_MBEDTLS when md4 available

---
 lib/curl_setup.h | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/lib/curl_setup.h b/lib/curl_setup.h
index 9f29b2e..5d82e33 100644
--- a/lib/curl_setup.h
+++ b/lib/curl_setup.h
@@ -634,11 +634,17 @@ int netware_init(void);
 /* Single point where USE_NTLM definition might be defined */
 #if !defined(CURL_DISABLE_NTLM) && !defined(CURL_DISABLE_CRYPTO_AUTH)
 #if defined(USE_OPENSSL) || defined(USE_WINDOWS_SSPI) || \
-    defined(USE_GNUTLS) || defined(USE_MBEDTLS) || defined(USE_NSS) || \
-    defined(USE_DARWINSSL) || defined(USE_OS400CRYPTO) || \
-    defined(USE_WIN32_CRYPTO)
+    defined(USE_GNUTLS) || defined(USE_NSS) || defined(USE_DARWINSSL) || \
+    defined(USE_OS400CRYPTO) || defined(USE_WIN32_CRYPTO)
 
 #define USE_NTLM
+
+#elif defined(USE_MBEDTLS)
+#  include <mbedtls/md4.h>
+#  if defined(MBEDTLS_MD4_C)
+#define USE_NTLM
+#  endif
+
 #endif
 #endif
 
