diff --git a/Makefile b/Makefile
index 5f4eb22..983f669 100644
--- a/Makefile
+++ b/Makefile
@@ -18,18 +18,30 @@
 # FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 # IN THE SOFTWARE.
 
+VERSION = 2.7.0
 PLATFORM ?= $(shell sh -c 'uname -s | tr "[A-Z]" "[a-z]"')
+ifeq ($(findstring cygwin,$(PLATFORM)), cygwin)
+PLATFORM = winnt
+endif
+ifeq ($(findstring mingw,$(PLATFORM)), mingw)
+PLATFORM = winnt
+endif
+
 HELPER ?=
 BINEXT ?=
 ifeq (darwin,$(PLATFORM))
-SONAME ?= libhttp_parser.2.7.0.dylib
+SONAME ?= libhttp_parser.$(VERSION).dylib
 SOEXT ?= dylib
+else ifeq (winnt, $(PLATFORM))
+CC = gcc
+SONAME ?= libhttp_parser.$(VERSION).dll
+SOEXT  ?= dll
 else ifeq (wine,$(PLATFORM))
 CC = winegcc
 BINEXT = .exe.so
 HELPER = wine
 else
-SONAME ?= libhttp_parser.so.2.7.0
+SONAME ?= libhttp_parser.so.$(VERSION)
 SOEXT ?= so
 endif
 
@@ -50,7 +62,12 @@ CFLAGS += -Wall -Wextra -Werror
 CFLAGS_DEBUG = $(CFLAGS) -O0 -g $(CFLAGS_DEBUG_EXTRA)
 CFLAGS_FAST = $(CFLAGS) -O3 $(CFLAGS_FAST_EXTRA)
 CFLAGS_BENCH = $(CFLAGS_FAST) -Wno-unused-parameter
+
+ifeq (winnt, $(PLATFORM))
+CFLAGS_LIB = $(CFLAGS_FAST)
+else
 CFLAGS_LIB = $(CFLAGS_FAST) -fPIC
+endif
 
 LDFLAGS_LIB = $(LDFLAGS) -shared
 
@@ -141,7 +158,7 @@ clean:
 	rm -f *.o *.a tags test test_fast test_g \
 		http_parser.tar libhttp_parser.so.* \
 		url_parser url_parser_g parsertrace parsertrace_g \
-		*.exe *.exe.so
+		*.exe *.exe.so *.dll
 
 contrib/url_parser.c:	http_parser.h
 contrib/parsertrace.c:	http_parser.h
