diff --git a/src/transports/ssh.c b/src/transports/ssh.c
index 89f085230..b8bdca61a 100644
--- a/src/transports/ssh.c
+++ b/src/transports/ssh.c
@@ -467,6 +467,7 @@ static int _git_ssh_setup_conn(
 	git_credential *cred = NULL;
 	LIBSSH2_SESSION *session=NULL;
 	LIBSSH2_CHANNEL *channel=NULL;
+	char *host_and_port;
 
 	t->current_stream = NULL;
 
@@ -567,10 +568,18 @@ static int _git_ssh_setup_conn(
 
 		cert_ptr = &cert;
 
+		if (atoi(s->url.port) == SSH_DEFAULT_PORT) {
+			host_and_port = s->url.host;
+		} else {
+			size_t n = strlen(s->url.host) + strlen(s->url.port) + 2;
+			host_and_port = alloca(n);
+			sprintf(host_and_port, "%s:%s", s->url.host, s->url.port);
+		}
+
 		error = t->owner->connect_opts.callbacks.certificate_check(
 			(git_cert *)cert_ptr,
 			0,
-			s->url.host,
+			host_and_port,
 			t->owner->connect_opts.callbacks.payload);
 
 		if (error < 0 && error != GIT_PASSTHROUGH) {
