From d64f3d0992ec278d843c397b4b52e3434962c197 Mon Sep 17 00:00:00 2001
From: Vinz2008 <68145293+Vinz2008@users.noreply.github.com>
Date: Thu, 11 Aug 2022 00:25:31 +0200
Subject: [PATCH] Fix #6365

---
 src/cli/opt.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/cli/opt.c b/src/cli/opt.c
index 72df5877fbf..62a3430d16e 100644
--- a/src/cli/opt.c
+++ b/src/cli/opt.c
@@ -23,7 +23,7 @@
 #include "opt.h"
 
 #ifdef _WIN32
-# include <Windows.h>
+# include <windows.h>
 #else
 # include <fcntl.h>
 # include <sys/ioctl.h>
