--- tests/CMakeLists.txt	2016-10-24 23:43:39.000000000 -0700
+++ tests/CMakeLists.txt	2016-11-15 20:44:32.132344000 -0800
@@ -47,6 +47,7 @@
 check_include_files(arpa/inet.h HAVE_ARPA_INET_H)
 check_include_files(windows.h HAVE_WINDOWS_H)
 check_include_files(winsock2.h HAVE_WINSOCK2_H)
+check_include_files(netinet/in.h HAVE_NETINET_IN_H)
 configure_file(
   "${CMAKE_CURRENT_SOURCE_DIR}/libssh2_config_cmake.h.in"
   "${CMAKE_CURRENT_BINARY_DIR}/libssh2_config.h")
