--- configure.orig	2023-06-04 05:19:04
+++ configure	2023-06-07 08:35:11
@@ -18117,7 +18117,7 @@
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-llzma  $LIBS"
+LIBS="-L${libdir} -l:liblzma.a $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -18148,7 +18148,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_lzma_lzma_mf_is_supported" >&5
 $as_echo "$ac_cv_lib_lzma_lzma_mf_is_supported" >&6; }
 if test "x$ac_cv_lib_lzma_lzma_mf_is_supported" = xyes; then :
-  LIBLZMA=-llzma
+  LIBLZMA="-L${libdir} -l:liblzma.a"
 
 $as_echo "#define HAVE_LZMA 1" >>confdefs.h
 
