From 1fa6efaa946243004c45be92e66b324dc980df7d Mon Sep 17 00:00:00 2001
From: Valentin Churavy <v.churavy@gmail.com>
Date: Thu, 17 Sep 2020 23:22:45 +0200
Subject: [PATCH] clang-sa can't determine that !RHS implies !LHS

---
 llvm/include/llvm/ADT/FunctionExtras.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/llvm/ADT/FunctionExtras.h b/include/llvm/ADT/FunctionExtras.h
index 121aa527a5d..b9b6d829b14 100644
--- a/include/llvm/ADT/FunctionExtras.h
+++ b/include/llvm/ADT/FunctionExtras.h
@@ -193,9 +193,11 @@ public:
     // Copy the callback and inline flag.
     CallbackAndInlineFlag = RHS.CallbackAndInlineFlag;
 
+#ifndef __clang_analyzer__
     // If the RHS is empty, just copying the above is sufficient.
     if (!RHS)
       return;
+#endif
 
     if (!isInlineStorage()) {
       // The out-of-line case is easiest to move.
-- 
2.28.0

