From b498303066a63a203d24f739b2d2e0e56dca70d1 Mon Sep 17 00:00:00 2001
From: serge-sans-paille <sguelton@redhat.com>
Date: Tue, 10 Nov 2020 14:55:25 +0100
Subject: [PATCH] [nfc] Fix missing include

---
 llvm/utils/benchmark/src/benchmark_register.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/llvm/utils/benchmark/src/benchmark_register.h b/llvm/utils/benchmark/src/benchmark_register.h
index 0705e219f2fa..4caa5ad4da07 100644
--- a/utils/benchmark/src/benchmark_register.h
+++ b/utils/benchmark/src/benchmark_register.h
@@ -1,6 +1,7 @@
 #ifndef BENCHMARK_REGISTER_H
 #define BENCHMARK_REGISTER_H

+#include <limits>
 #include <vector>

 #include "check.h"
