diff --git a/Makefile.rules b/Makefile.rules
index 24cac3b..4f52dbb 100644
--- a/Makefile.rules
+++ b/Makefile.rules
@@ -1234,6 +1234,9 @@ endif
 DestSharedLib := $(DestSharedLibDir)/$(BaseLibName.SO)
 DestSharedAlias := $(DestSharedLibDir)/$(BaseAliasName.SO)
 
+$(DestSharedLibDir)::
+	$(MKDIR) $(DestSharedLibDir)
+
 install-local:: $(DestSharedLib)
 
 $(DestSharedLib): $(LibName.SO) $(DestSharedLibDir)
