diff -u llvm-3.9.1.orig/lib/Target/NVPTX/NVPTXISelLowering.cpp llvm-3.9.1/lib/Target/NVPTX/NVPTXISelLowering.cpp
--- llvm-3.9.1.orig/lib/Target/NVPTX/NVPTXISelLowering.cpp	2016-07-15 20:27:10.000000000 +0200
+++ llvm-3.9.1/lib/Target/NVPTX/NVPTXISelLowering.cpp	2017-09-07 04:04:25.540721694 +0200
@@ -888,6 +910,14 @@
   return TargetLoweringBase::getPreferredVectorAction(VT);
 }
 
+bool NVPTXTargetLowering::isNoopAddrSpaceCast(unsigned SrcAS,
+                                              unsigned DestAS) const {
+  assert(SrcAS != DestAS && "Expected different address spaces!");
+
+  return (SrcAS  == ADDRESS_SPACE_GENERIC || SrcAS  > ADDRESS_SPACE_LOCAL) &&
+         (DestAS == ADDRESS_SPACE_GENERIC || DestAS > ADDRESS_SPACE_LOCAL);
+}
+
 SDValue
 NVPTXTargetLowering::LowerGlobalAddress(SDValue Op, SelectionDAG &DAG) const {
   SDLoc dl(Op);
diff -u llvm-3.9.1.orig/lib/Target/NVPTX/NVPTXISelLowering.h llvm-3.9.1/lib/Target/NVPTX/NVPTXISelLowering.h
--- llvm-3.9.1.orig/lib/Target/NVPTX/NVPTXISelLowering.h	2016-06-12 17:39:02.000000000 +0200
+++ llvm-3.9.1/lib/Target/NVPTX/NVPTXISelLowering.h	2017-09-07 04:03:39.591373952 +0200
@@ -442,6 +442,8 @@
                                const NVPTXSubtarget &STI);
   SDValue LowerOperation(SDValue Op, SelectionDAG &DAG) const override;
 
+  bool isNoopAddrSpaceCast(unsigned SrcAS, unsigned DestAS) const override;
+
   SDValue LowerGlobalAddress(SDValue Op, SelectionDAG &DAG) const;
 
   const char *getTargetNodeName(unsigned Opcode) const override;
