diff -ru llvm-5.0.0.orig/lib/Target/NVPTX/NVPTXISelLowering.cpp llvm-5.0.0/lib/Target/NVPTX/NVPTXISelLowering.cpp
--- llvm-5.0.0.orig/lib/Target/NVPTX/NVPTXISelLowering.cpp	2017-07-12 22:49:21.000000000 +0200
+++ llvm-5.0.0/lib/Target/NVPTX/NVPTXISelLowering.cpp	2018-01-13 09:08:17.275987874 +0100
@@ -1235,6 +1235,14 @@
   }
 }
 
+bool NVPTXTargetLowering::isNoopAddrSpaceCast(unsigned SrcAS,
+                                              unsigned DestAS) const {
+  assert(SrcAS != DestAS && "Expected different address spaces!");
+
+  return (SrcAS  == ADDRESS_SPACE_GENERIC || SrcAS  > ADDRESS_SPACE_LOCAL) &&
+         (DestAS == ADDRESS_SPACE_GENERIC || DestAS > ADDRESS_SPACE_LOCAL);
+}
+
 SDValue
 NVPTXTargetLowering::LowerGlobalAddress(SDValue Op, SelectionDAG &DAG) const {
   SDLoc dl(Op);
diff -ru llvm-5.0.0.orig/lib/Target/NVPTX/NVPTXISelLowering.h llvm-5.0.0/lib/Target/NVPTX/NVPTXISelLowering.h
--- llvm-5.0.0.orig/lib/Target/NVPTX/NVPTXISelLowering.h	2018-01-13 09:07:48.839643576 +0100
+++ llvm-5.0.0/lib/Target/NVPTX/NVPTXISelLowering.h	2018-01-13 09:06:18.658551692 +0100
@@ -443,6 +443,8 @@
                                const NVPTXSubtarget &STI);
   SDValue LowerOperation(SDValue Op, SelectionDAG &DAG) const override;
 
+  bool isNoopAddrSpaceCast(unsigned SrcAS, unsigned DestAS) const override;
+
   SDValue LowerGlobalAddress(SDValue Op, SelectionDAG &DAG) const;
 
   const char *getTargetNodeName(unsigned Opcode) const override;
