diff --git a/tools/llvm-c-test/Makefile b/tools/llvm-c-test/Makefile
index 08be7c3..f56e83c 100644
--- a/tools/llvm-c-test/Makefile
+++ b/tools/llvm-c-test/Makefile
@@ -25,5 +25,5 @@ CFLAGS += -std=c99 -Wall -Wstrict-prototypes
 
 # ...but if it is built - use it
 ifeq ($(ENABLE_SHARED),1)
-LIBS = -lLLVM-$(LLVMVersion)
+LIBS = -lLLVM-$(LLVM_VERSION_MAJOR).$(LLVM_VERSION_MINOR)$(LLVM_VERSION_SUFFIX)
 endif
diff --git a/tools/llvm-shlib/Makefile b/tools/llvm-shlib/Makefile
index 2bc81da..3f7652f 100644
--- a/tools/llvm-shlib/Makefile
+++ b/tools/llvm-shlib/Makefile
@@ -15,7 +15,6 @@ LIBRARYALIASNAME = LLVM-$(LLVMVersion)
 NO_BUILD_ARCHIVE := 1
 LINK_LIBS_IN_SHARED := 1
 SHARED_LIBRARY := 1
-SHARED_ALIAS := 1
 
 include $(LEVEL)/Makefile.config
 
@@ -27,6 +26,8 @@ ifeq ($(HOST_OS), $(filter $(HOST_OS), Cygwin MingW))
     SHLIB_FRAG_NAMES += stdc++.a.o
   endif
 
+else
+  SHARED_ALIAS := 1
 endif
 
 include $(LEVEL)/Makefile.common
