--- a/projects/compiler-rt/cmake/builtin-config-ix.cmake
+++ b/projects/compiler-rt/cmake/builtin-config-ix.cmake
@@ -57,9 +57,6 @@
   find_darwin_sdk_dir(DARWIN_tvos_SYSROOT appletvos)
 
   set(DARWIN_EMBEDDED_PLATFORMS)
-  set(DARWIN_osx_BUILTIN_MIN_VER 10.5)
-  set(DARWIN_osx_BUILTIN_MIN_VER_FLAG
-      -mmacosx-version-min=${DARWIN_osx_BUILTIN_MIN_VER})
 
   if(COMPILER_RT_ENABLE_IOS)
     list(APPEND DARWIN_EMBEDDED_PLATFORMS ios)
@@ -101,20 +98,6 @@
     set(CAN_TARGET_${arch} 1)
   endforeach()
 
-  # Need to build a 10.4 compatible libclang_rt
-  set(DARWIN_10.4_SYSROOT ${DARWIN_osx_SYSROOT})
-  set(DARWIN_10.4_BUILTIN_MIN_VER 10.4)
-  set(DARWIN_10.4_BUILTIN_MIN_VER_FLAG
-      -mmacosx-version-min=${DARWIN_10.4_BUILTIN_MIN_VER})
-  set(DARWIN_10.4_SKIP_CC_KEXT On)
-  darwin_test_archs(10.4 DARWIN_10.4_ARCHS i386 x86_64)
-  message(STATUS "OSX 10.4 supported builtin arches: ${DARWIN_10.4_ARCHS}")
-  if(DARWIN_10.4_ARCHS)
-    # don't include the Haswell slice in the 10.4 compatibility library
-    list(REMOVE_ITEM DARWIN_10.4_ARCHS x86_64h)
-    list(APPEND BUILTIN_SUPPORTED_OS 10.4)
-  endif()
-
   foreach(platform ${DARWIN_EMBEDDED_PLATFORMS})
     if(DARWIN_${platform}sim_SYSROOT)
       set(DARWIN_${platform}sim_BUILTIN_MIN_VER
