--- driver/others/blas_server.c	2016-08-31 20:58:42.000000000 -0700
+++ driver/others/blas_server.c	2016-11-16 21:24:52.282666000 -0800
@@ -70,7 +70,7 @@
 /*********************************************************************/
 
 #include "common.h"
-#if defined(OS_LINUX) || defined(OS_NETBSD) || defined(OS_DARWIN) || defined(OS_ANDROID) || defined(OS_SUNOS)
+#if defined(OS_LINUX) || defined(OS_NETBSD) || defined(OS_DARWIN) || defined(OS_ANDROID) || defined(OS_SUNOS) || defined(OS_FREEBSD)
 #include <dlfcn.h>
 #include <signal.h>
 #include <sys/resource.h>
