diff --git a/src/pcre2grep.c b/src/pcre2grep.c
index e98d743..f00dac2 100644
--- a/src/pcre2grep.c
+++ b/src/pcre2grep.c
@@ -628,7 +628,7 @@ z/OS, and "no support". */
 
 /************* Directory scanning Unix-style and z/OS ***********/
 
-#if (defined HAVE_SYS_STAT_H && defined HAVE_DIRENT_H && defined HAVE_SYS_TYPES_H) || defined NATIVE_ZOS
+#if ((defined HAVE_SYS_STAT_H && defined HAVE_DIRENT_H && defined HAVE_SYS_TYPES_H) || defined NATIVE_ZOS) && !defined WIN32
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <dirent.h>
